akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -39,92 +125,85 @@ import scala.util.control.NonFatal
   extends AbstractExecutorService {
 
   if (parallelism <= 0)
     throw new IllegalArgumentException("Size of pool cannot be less or equal to 0")
 
+  import AffinityPool._
+
   // Held while starting/shutting down workers/pool in order to make
   // the operations linear and enforce atomicity. An example of that would be
   // adding a worker. We want the creation of the worker, addition
   // to the set and starting to worker to be an atomic action. Using
   // a concurrent set would not give us that
-  private val bookKeepingLock = new ReentrantLock()
+  private val bookKeepingLock = new ReentrantGuard()
 
   // condition used for awaiting termination
   private val terminationCondition = bookKeepingLock.newCondition()
 
   // indicates the current state of the pool
-  @volatile final private var poolState: PoolState = Running
+  @volatile final private var poolState: PoolState = Uninitialized
 
-  private final val workQueues = Array.fill(parallelism)(new BoundedTaskQueue(affinityGroupSize))
-  private final val workers = mutable.Set[ThreadPoolWorker]()
+  private final val workQueues = Array.fill(parallelism)(new BoundedAffinityTaskQueue(affinityGroupSize))
+  private final val workers = mutable.Set[AffinityPoolWorker]()
 
-  // a counter that gets incremented every time a task is queued
-  private val executionCounter: AtomicInteger = new AtomicInteger(0)
   // maps a runnable to an index of a worker queue
-  private val runnableToWorkerQueueIndex = new AtomicReference(ImmutableIntMap.empty)
-
-  private def locked[T](l: Lock)(body: ⇒ T) = {
-    l.lock()
-    try {
-      body
-    } finally {
-      l.unlock()
-    }
-  }
-
-  private def getQueueForRunnable(command: Runnable) = {
+  private[this] final val hashCache = new AtomicReference(ImmutableIntMap.empty)
 
+  private def getQueueForRunnable(command: Runnable): BoundedAffinityTaskQueue = {
     val runnableHash = command.hashCode()
 
-    def sbhash(i: Int) = reverseBytes(i * 0x9e3775cd) * 0x9e3775cd
-
-    def getNext = executionCounter.incrementAndGet() % parallelism
+    def indexFor(h: Int): Int =
+      Math.abs(reverseBytes(h * 0x9e3775cd) * 0x9e3775cd) % parallelism // In memory of Phil Bagwell
 
-    def updateIfAbsentAndGetQueueIndex(
-      workerQueueIndex: AtomicReference[ImmutableIntMap],
-      runnableHash:     Int, queueIndex: ⇒ Int): Int = {
-      @tailrec
-      def updateIndex(): Unit = {
-        val prev = workerQueueIndex.get()
-        if (!runnableToWorkerQueueIndex.compareAndSet(prev, prev.updateIfAbsent(runnableHash, queueIndex))) {
-          updateIndex()
+    val workQueueIndex =
+      if (fairDistributionThreshold == 0)
+        indexFor(runnableHash)
+      else {
+        @tailrec
+        def cacheLookup(prev: ImmutableIntMap, hash: Int): Int = {
+          val existingIndex = prev.get(runnableHash)
+          if (existingIndex >= 0) existingIndex
+          else if (prev.size > fairDistributionThreshold) indexFor(hash)
```

Comment:
This consults the map regardless of whether we have already crossed the `fairDistributionThreshold`. If we disregard the cache entirely after hitting the `fairDistributionThreshold` we incur a performance hit every time we cross the boundary, but are faster when we're over it.

Tried running a benchmark, appears to make a nice difference. How do you generate those bar charts?

Before:
```
[info] Benchmark                                                            (dispatcher)                           (mailbox)  (throughPut)   Mode  Cnt        Score          Error  Units
[info] AffinityPoolRequestResponseBenchmark.queryUserServiceActor    affinity-dispatcher  SingleConsumerOnlyUnboundedMailbox             1  thrpt    3  3042671.765 ±  6864270.945  ops/s
[info] AffinityPoolRequestResponseBenchmark.queryUserServiceActor    affinity-dispatcher  SingleConsumerOnlyUnboundedMailbox             5  thrpt    3  3743330.212 ± 23215318.467  ops/s
[info] AffinityPoolRequestResponseBenchmark.queryUserServiceActor    affinity-dispatcher  SingleConsumerOnlyUnboundedMailbox            50  thrpt    3  3785579.970 ± 15947684.518  ops/s
```

After:
```
[info] AffinityPoolRequestResponseBenchmark.queryUserServiceActor    affinity-dispatcher  SingleConsumerOnlyUnboundedMailbox             1  thrpt    3  3629568.121 ±  7889401.954  ops/s
[info] AffinityPoolRequestResponseBenchmark.queryUserServiceActor    affinity-dispatcher  SingleConsumerOnlyUnboundedMailbox             5  thrpt    3  4483726.757 ± 12557212.880  ops/s
[info] AffinityPoolRequestResponseBenchmark.queryUserServiceActor    affinity-dispatcher  SingleConsumerOnlyUnboundedMailbox            50  thrpt    3  4764827.280 ± 12213405.339  ops/s
```